explorer/ast/value_node.h

```diff
@@ -0,0 +1,143 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_AST_VALUE_NODE_H_
+#define CARBON_EXPLORER_AST_VALUE_NODE_H_
+
+#include <functional>
+#include <optional>
+#include <string_view>
+
+#include "explorer/ast/ast_node.h"
+#include "explorer/ast/value_category.h"
+#include "explorer/common/nonnull.h"
+
+namespace Carbon {
+
+class Value;
+
+// The placeholder name exposed by anonymous ValueNodes.
+static constexpr std::string_view AnonymousName = "_";
+
+// ImplementsValueNode is true if NodeType::ImplementsCarbonValueNode
+// is valid and names a type, indicating that NodeType implements the
+// ValueNode interface, defined below.
+
+template <typename NodeType, typename = void>
+static constexpr bool ImplementsValueNode = false;
+
+// ValueNode is an interface implemented by AstNodes that can be associated
+// with a value, such as declarations and bindings. The interface consists of
+// the following methods:
+//
+// // Returns the constant associated with the node.
+// // This is called by the interpreter, not the type checker.
+// auto constant_value() const -> std::optional<Nonnull<const Value*>>;
+//
+// // Returns the symbolic compile-time identity of the node.
+// // This is called by the type checker, not the interpreter.
+// auto symbolic_identity() const -> std::optional<Nonnull<const Value*>>;
+//
+// // Returns the static type of an IdentifierExpression that names *this.
+// auto static_type() const -> const Value&;
+//
+// // Returns the value category of an IdentifierExpression that names *this.
+// auto value_category() const -> ValueCategory;
+//
+// // Print the node's identity (e.g. its name).
+// void PrintID(llvm::raw_ostream& out) const;
+//
+// TODO: consider turning the above documentation into real code, as sketched
+// at https://godbolt.org/z/186oEozhc
+
+template <typename T>
+static constexpr bool
+    ImplementsValueNode<T, typename T::ImplementsCarbonValueNode> = true;
+
+class ValueNodeView {
```

Comment:
I'd suggest value_node_view.h to better match the class name

Reply:
This file also defines the `ValueNode` concept (albeit with words rather than code for now) and `ImplementsValueNode`.